# Discogs Analysis


[Discogs Developer](https://www.discogs.com/developers/#page:home,header:home-quickstart)

[Download Discogs Data](https://data.discogs.com/)

[Discogs Database Search](https://www.discogs.com/developers/#page:database,header:database-search)

[Discogs Oauth](https://github.com/jesseward/discogs-oauth-example)

Need data on location of purchase

### Key Questions to answer:

* What is the profile of my record collection?

* What are the most expensive records and CDs I own?

* Which band/artist do I own the most and what is the cost of that?

* Which record label do I support the most and which did I spent the most?

* Which subgenre costs the most and which one is worth the most?

* Is there a difference between autographed and non-authographed records?

* What is the breakdown of the format and the cost, min price, median price, and max price?

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt 
import seaborn as sns
import ydata-profiling
#import cufflinks as cf
import plotly.offline
#cf.go_offline()
#cf.set_config_file(offline=False, world_readable=True)
from pylab import rcParams
sns.set(rc={'figure.figsize':(16,12)})
plt.style.use('ggplot')
pd.set_option('display.max_columns', None)
pd.set_option('display.max_row', None)

In [3]:
discogs_df = pd.read_csv('../data/discogs_data.csv')
discogs_df.head(10)

,ID,Catalog#,Artist,Artist2,Title,Label,Format,Rating,Released,release_id,CollectionFolder,Date Added,Collection Media Condition,Collection Sleeve Condition,Collection Notes,Collection Cost,Collection Autographed,Min,Median,Max,Genre,Subgenre
0,33,SD 8283,Yes,Yes,The Yes Album,Atlantic,"LP, Album, RE, PR;",NaN,1971,7116823.0,Vinyl,7/21/19 20:21,NaN,NaN,NaN,5.50,NaN,2.25,3.87,10.00,Rock,Progressive Rock
1,165,RR 6583-1,Mastodon,Mastodon,Remission,Relapse Records,"2x12"", Album, RE, RM, RP",NaN,2017,8863478.0,Vinyl,12/3/16 11:44,NaN,NaN,NaN,15.00,NaN,7.99,21.67,30.19,Metal,Sludge Metal
2,373,CS 8192,The Dave Brubeck Quartet,The Dave Brubeck Quartet,Time Out,Columbia,"LP, Album",NaN,1961,2676016.0,Vinyl,5/5/14 21:16,NaN,NaN,NaN,20.00,NaN,5.07,26.50,130.00,Jazz,Jazz
3,142,SD 3-100,Yes,Yes,Yessongs,Atlantic,"3xLP, Album",5.0,1973,8179158.0,Vinyl,4/20/17 19:05,NaN,NaN,NaN,5.00,NaN,3.95,9.99,20.00,Rock,Progressive Rock
4,18,SD 2-908,Yes,Yes,Tales From Topographic Oceans,Atlantic,"2xLP, Album, Gat",NaN,1973,8535940.0,Vinyl,2/22/20 15:50,NaN,NaN,NaN,5.50,NaN,7.99,13.49,20.00,Rock,Progressive Rock
5,340,SD 18122,Yes,Yes,Relayer,Atlantic,"LP, Album, Gat",NaN,1974,755646.0,Vinyl,9/6/14 11:59,NaN,NaN,NaN,4.00,NaN,1.13,7.00,30.74,Rock,Progressive Rock
6,374,SD 8245,King Crimson,King Crimson,In The Court Of The Crimson King An Observati...,Atlantic,"LP, Album, RE, PRC",NaN,1975,4254321.0,Vinyl,5/5/14 21:14,NaN,NaN,NaN,12.00,NaN,10.00,17.00,25.00,Rock,Progressive Rock
7,255,"34224, PZ 34224",Kansas (2),Kansas (2),Leftoverture,"Kirshner, Kirshner","LP, Album",3.0,1976,1476710.0,Vinyl,8/9/15 18:25,NaN,NaN,NaN,0.86,NaN,1.72,3.82,9.95,Rock,Classic Rock
8,341,SD 19132,Yes,Yes,Fragile,Atlantic,"LP, Album, RE, SRC",NaN,1977,2637700.0,Vinyl,9/6/14 11:59,NaN,NaN,NaN,4.00,NaN,4.50,7.25,14.00,Rock,Progressive Rock
9,19,AFL1-3516,Scorpions,Scorpions,Best Of Scorpions,RCA Victor,"LP, Comp",NaN,1979,3632339.0,Vinyl,2/22/20 15:47,NaN,NaN,NaN,2.00,NaN,4.50,7.99,14.99,Metal,Heavy Metal


Some of the labels are separated by commas.

In [4]:
discogs_df = pd.concat([discogs_df, discogs_df['Label'].str.split(', ', expand=True)], axis=1)
discogs_df.head(20)

,ID,Catalog#,Artist,Artist2,Title,Label,Format,Rating,Released,release_id,CollectionFolder,Date Added,Collection Media Condition,Collection Sleeve Condition,Collection Notes,Collection Cost,Collection Autographed,Min,Median,Max,Genre,Subgenre,0,1,2,3
0,33,SD 8283,Yes,Yes,The Yes Album,Atlantic,"LP, Album, RE, PR;",NaN,1971,7116823.0,Vinyl,7/21/19 20:21,NaN,NaN,NaN,5.50,NaN,2.25,3.87,10.00,Rock,Progressive Rock,Atlantic,NaN,NaN,NaN
1,165,RR 6583-1,Mastodon,Mastodon,Remission,Relapse Records,"2x12"", Album, RE, RM, RP",NaN,2017,8863478.0,Vinyl,12/3/16 11:44,NaN,NaN,NaN,15.00,NaN,7.99,21.67,30.19,Metal,Sludge Metal,Relapse Records,NaN,NaN,NaN
2,373,CS 8192,The Dave Brubeck Quartet,The Dave Brubeck Quartet,Time Out,Columbia,"LP, Album",NaN,1961,2676016.0,Vinyl,5/5/14 21:16,NaN,NaN,NaN,20.00,NaN,5.07,26.50,130.00,Jazz,Jazz,Columbia,NaN,NaN,NaN
3,142,SD 3-100,Yes,Yes,Yessongs,Atlantic,"3xLP, Album",5.0,1973,8179158.0,Vinyl,4/20/17 19:05,NaN,NaN,NaN,5.00,NaN,3.95,9.99,20.00,Rock,Progressive Rock,Atlantic,NaN,NaN,NaN
4,18,SD 2-908,Yes,Yes,Tales From Topographic Oceans,Atlantic,"2xLP, Album, Gat",NaN,1973,8535940.0,Vinyl,2/22/20 15:50,NaN,NaN,NaN,5.50,NaN,7.99,13.49,20.00,Rock,Progressive Rock,Atlantic,NaN,NaN,NaN
5,340,SD 18122,Yes,Yes,Relayer,Atlantic,"LP, Album, Gat",NaN,1974,755646.0,Vinyl,9/6/14 11:59,NaN,NaN,NaN,4.00,NaN,1.13,7.00,30.74,Rock,Progressive Rock,Atlantic,NaN,NaN,NaN
6,374,SD 8245,King Crimson,King Crimson,In The Court Of The Crimson King An Observati...,Atlantic,"LP, Album, RE, PRC",NaN,1975,4254321.0,Vinyl,5/5/14 21:14,NaN,NaN,NaN,12.00,NaN,10.00,17.00,25.00,Rock,Progressive Rock,Atlantic,NaN,NaN,NaN
7,255,"34224, PZ 34224",Kansas (2),Kansas (2),Leftoverture,"Kirshner, Kirshner","LP, Album",3.0,1976,1476710.0,Vinyl,8/9/15 18:25,NaN,NaN,NaN,0.86,NaN,1.72,3.82,9.95,Rock,Classic Rock,Kirshner,Kirshner,NaN,NaN
8,341,SD 19132,Yes,Yes,Fragile,Atlantic,"LP, Album, RE, SRC",NaN,1977,2637700.0,Vinyl,9/6/14 11:59,NaN,NaN,NaN,4.00,NaN,4.50,7.25,14.00,Rock,Progressive Rock,Atlantic,NaN,NaN,NaN
9,19,AFL1-3516,Scorpions,Scorpions,Best Of Scorpions,RCA Victor,"LP, Comp",NaN,1979,3632339.0,Vinyl,2/22/20 15:47,NaN,NaN,NaN,2.00,NaN,4.50,7.99,14.99,Metal,Heavy Metal,RCA Victor,NaN,NaN,NaN


Rename columns

In [5]:
discogs_df.rename(columns={0:'Label1',1:'Label2',2:'Label3',3:'Label4'}, inplace=True)
discogs_df.head()

,ID,Catalog#,Artist,Artist2,Title,Label,Format,Rating,Released,release_id,CollectionFolder,Date Added,Collection Media Condition,Collection Sleeve Condition,Collection Notes,Collection Cost,Collection Autographed,Min,Median,Max,Genre,Subgenre,Label1,Label2,Label3,Label4
0,33,SD 8283,Yes,Yes,The Yes Album,Atlantic,"LP, Album, RE, PR;",NaN,1971,7116823.0,Vinyl,7/21/19 20:21,NaN,NaN,NaN,5.5,NaN,2.25,3.87,10.00,Rock,Progressive Rock,Atlantic,NaN,NaN,NaN
1,165,RR 6583-1,Mastodon,Mastodon,Remission,Relapse Records,"2x12"", Album, RE, RM, RP",NaN,2017,8863478.0,Vinyl,12/3/16 11:44,NaN,NaN,NaN,15.0,NaN,7.99,21.67,30.19,Metal,Sludge Metal,Relapse Records,NaN,NaN,NaN
2,373,CS 8192,The Dave Brubeck Quartet,The Dave Brubeck Quartet,Time Out,Columbia,"LP, Album",NaN,1961,2676016.0,Vinyl,5/5/14 21:16,NaN,NaN,NaN,20.0,NaN,5.07,26.50,130.00,Jazz,Jazz,Columbia,NaN,NaN,NaN
3,142,SD 3-100,Yes,Yes,Yessongs,Atlantic,"3xLP, Album",5.0,1973,8179158.0,Vinyl,4/20/17 19:05,NaN,NaN,NaN,5.0,NaN,3.95,9.99,20.00,Rock,Progressive Rock,Atlantic,NaN,NaN,NaN
4,18,SD 2-908,Yes,Yes,Tales From Topographic Oceans,Atlantic,"2xLP, Album, Gat",NaN,1973,8535940.0,Vinyl,2/22/20 15:50,NaN,NaN,NaN,5.5,NaN,7.99,13.49,20.00,Rock,Progressive Rock,Atlantic,NaN,NaN,NaN


Export for inspection

In [6]:
discogs_df.to_csv('../output/discogs_clean_data.csv',index=False)

## Label Sorting

Label must be sorted. There is an issue like this where Svart Records is placed before 20 Buck Spin

In [7]:
twentybuck = discogs_df[discogs_df['Label'].str.contains("20 Buck Spin")]
twentybuck

,ID,Catalog#,Artist,Artist2,Title,Label,Format,Rating,Released,release_id,CollectionFolder,Date Added,Collection Media Condition,Collection Sleeve Condition,Collection Notes,Collection Cost,Collection Autographed,Min,Median,Max,Genre,Subgenre,Label1,Label2,Label3,Label4
154,88,SPIN065,Mournful Congregation,Mournful Congregation,Concrescence Of The Sophia,20 Buck Spin,"LP, MiniAlbum",NaN,2014,5807031.0,Vinyl,6/3/18 15:21,NaN,NaN,NaN,17.0,NaN,6.00,9.00,15.99,Metal,Doom Metal,20 Buck Spin,NaN,NaN,NaN
215,156,SPIN075,Khemmis,Khemmis,Absolution,20 Buck Spin,"LP, Album, RP",NaN,2015,8551023.0,Vinyl,1/18/17 15:05,NaN,NaN,NaN,18.0,Yes,10.00,19.49,24.99,Metal,Doom Metal,20 Buck Spin,NaN,NaN,NaN
239,213,SPIN048,Pallbearer,Pallbearer,Sorrow And Extinction,20 Buck Spin,"2xLP, Album, RP, Blu",4.0,2015,6825240.0,Vinyl,2/26/16 11:05,NaN,NaN,NaN,15.0,NaN,20.00,29.43,36.30,Metal,Doom Metal,20 Buck Spin,NaN,NaN,NaN
259,157,SPIN076,Khemmis,Khemmis,Hunted,20 Buck Spin,"LP, Album",NaN,2016,9222656.0,Vinyl,1/18/17 15:05,NaN,NaN,NaN,18.0,Yes,10.00,15.00,23.33,Metal,Doom Metal,20 Buck Spin,NaN,NaN,NaN
283,14,"SVART003, SPIN 080",Oranssi Pazuzu,Oranssi Pazuzu,Värähtelijä,"Svart Records, 20 Buck Spin","2xLP, Ltd, Blu",NaN,2016,11636087.0,Vinyl,2/22/20 15:59,NaN,NaN,NaN,25.0,NaN,20.00,24.17,26.37,Metal,Black Metal,Svart Records,20 Buck Spin,NaN,NaN
285,133,SPIN072,Obsequiae,Obsequiae,Aria Of Vernal Tombs,20 Buck Spin,"LP, Album, Cle",NaN,2016,8556349.0,Vinyl,4/27/17 18:58,NaN,NaN,NaN,15.0,NaN,20.78,28.50,36.66,Metal,Black Metal,20 Buck Spin,NaN,NaN,NaN
311,91,SPIN091,Acephalix,Acephalix,Decreation,20 Buck Spin,"LP, Album",NaN,2017,10946608.0,Vinyl,5/29/18 18:20,NaN,NaN,NaN,15.0,NaN,11.90,18.50,27.47,Metal,Death Metal,20 Buck Spin,NaN,NaN,NaN
322,71,"SVR226, SPIN059",Oranssi Pazuzu,Oranssi Pazuzu,Valonielu,"Svart Records, 20 Buck Spin","LP, Album, RE",NaN,2017,10341101.0,Vinyl,12/18/18 14:36,NaN,NaN,NaN,15.0,NaN,18.71,22.06,26.56,Metal,Black Metal,Svart Records,20 Buck Spin,NaN,NaN
346,73,SPIN093,Spirit Adrift,Spirit Adrift,Curse Of Conception,20 Buck Spin,"LP, Album, Oli",NaN,2018,12305606.0,Vinyl,12/18/18 12:39,NaN,NaN,NaN,15.0,NaN,12.00,21.00,23.99,Metal,Doom Metal,20 Buck Spin,NaN,NaN,NaN
350,79,"SPIN081, OPLP338",Mournful Congregation,Mournful Congregation,The Incubus Of Karma,"20 Buck Spin, Osmose Productions","2xLP, Album, Ltd",5.0,2018,11750808.0,Vinyl,8/4/18 8:30,NaN,NaN,NaN,27.0,Yes,14.00,20.32,23.91,Metal,Doom Metal,20 Buck Spin,Osmose Productions,NaN,NaN


In [8]:
discogs_df['Label'][discogs_df['Artist'] == 'Oranssi Pazuzu']

283    Svart Records, 20 Buck Spin
322    Svart Records, 20 Buck Spin
393               Roadburn Records
Name: Label, dtype: str

In [9]:
list = ['Svart Records', '20 Buck Spin']
list

['Svart Records', '20 Buck Spin']

Sort like this

In [10]:
list.sort()
list

['20 Buck Spin', 'Svart Records']

Drop column

In [11]:
# Drop columns
discogs_df = discogs_df.drop(columns=['Rating','Collection Media Condition', 
                                      'Collection Notes','Collection Sleeve Condition'])

In [12]:
discogs_df['Collection Autographed'] = discogs_df['Collection Autographed'].fillna('No')

## Create new columns:

* Artist - Label  
* pct_profit_min  
* pct_profit_median  
* pct_profit_max  

In [13]:
discogs_df['Record'] = discogs_df['Artist'] + ' - ' + discogs_df['Title']
discogs_df[['Artist','Label','Record']].head()

,Artist,Label,Record
0,Yes,Atlantic,Yes - The Yes Album
1,Mastodon,Relapse Records,Mastodon - Remission
2,The Dave Brubeck Quartet,Columbia,The Dave Brubeck Quartet - Time Out
3,Yes,Atlantic,Yes - Yessongs
4,Yes,Atlantic,Yes - Tales From Topographic Oceans


Run Pandas Profiler

In [ ]:
profile = discogs_df.profile_report(title='Discogs Data Profiling Report',correlations={"cramers": False})

In [ ]:
profile

Here are some interesting information about my record collection: 

1) I purchased around 408 albums in vinyl, CD, and tape format.  
2) I have 188 unique artists in my collection. I also have 10 artists that collaborated with others.  
3) My top five bands are:

|        Artist2              |   Count   |  Percentage  |
|:---------------------------:|:---------:|:------------:|
|  Opeth	                  |    19	  |      4.7%	 |
|  Cult Of Luna	              |    16	  |      3.9%	 |
|  Isis (6)	                  |    12	  |      2.9%	 |
|  Panopticon (6)	          |    11	  |      2.7%	 |
|  Between The Buried And Me  |    10	  |      2.5%	 |

4) I own 388 distinct albums.   
5) In my collection, 84.3% are vinyl and 15.4\% are CD. I have 2 Tapes.

6) My top 5 genres are:     

|  Genre       |  Count  |  Percentage  |
|:------------:|:-------:|:------------:|
|  Metal	   |   358	 |     87.7%    |	 
|  Rock	       |   31	 |     7.6%	    | 
|  Electronic  |   12	 |     2.9%	    | 
|  Pop	       |   3	 |     0.7%	    | 
|  Hip Hop	   |   2	 |     0.5%	    |

7) My top 5 subgenres are:  

|       Genre          |  Count  |  Percentage  |
|:--------------------:|:-------:|:------------:|
|  Black Metal	       |    83	 |     20.3%    | 
|  Progressive Metal   |    62	 |     15.2%    | 
|  Post Metal	       |    61	 |     15.0%    | 
|  Djent	           |    26	 |     6.4%	    |
|  Sludge Metal	       |    26	 |     6.4%	    |


7) I typically pay \\$15 (50 or 12.3\%) for a record. The next is \\$20 (49 or 12.0\%).     
8) The most frequent album sold at the `_Max_` price is \\$29.99 (12 or 2.9\%). Next is \\$39.99 (9 or 2.2\%).

9) My top five record labels are:  

|        Label         |  Count  |  Percentage  |
|:--------------------:|:-------:|:------------:|
|  Relapse Records	   |    23	 |      5.6%	| 
|  Century Media	   |    17	 |      4.2%	| 
|  Prosthetic Records  |    13	 |      3.2%	| 
|  Hydra Head Records  |    12	 |      2.9%	| 
|  20 Buck Spin	       |    10	 |      2.5%	|

10) Most albums I buy were released in the last several years: 

|  Year  |  Count  |  Percentage  |
|:------:|:-------:|:------------:|
|  2014	 |   61	   |     15.0%    |	 
|  2013	 |   55	   |     13.5%    |	 
|  2016	 |   52	   |     12.7%    |	 
|  2015	 |   48	   |     11.8%    |
|  2017	 |   40	   |     9.8%     |

Get summary statistics

In [ ]:
discogs_df.describe()

Run correlation

In [ ]:
corrMatrix = discogs_df.corr()
sns.heatmap(corrMatrix, annot=True)

## Plot boxplot of the collection cost, minimum sold, median sold, and max sold.

In [ ]:
sns.boxplot(x="variable", y="value", data=pd.melt(discogs_df[['Collection_Cost','_Min_','_Median_','_Max_']]))

Remove outlier

In [ ]:
sns.boxplot(x="variable", y="value", data=pd.melt(discogs_df[['Collection_Cost','_Min_','_Median_','_Max_']][discogs_df.Collection_Cost < discogs_df['Collection_Cost'].max()]))

In [ ]:
sns.boxplot(x="variable", y="value", data=pd.melt(discogs_df[['Collection_Cost','_Min_','_Median_']][discogs_df.Collection_Cost < discogs_df['Collection_Cost'].max()]))

In [ ]:
discogs_df[['Collection_Cost','_Min_','_Median_','_Max_']].sum().reset_index()

In [ ]:
print('I spent $', discogs_df['Collection_Cost'].sum(),' on music that is recorded on Discogs. \n \nMy collection is worth $',discogs_df['_Median_'].sum())

In [ ]:
discogs_df[['Collection_Cost','_Min_','_Median_','_Max_']].mean().reset_index()

In [ ]:
print('I spent on averege $', discogs_df['Collection_Cost'].mean().round(2),' on per album that is recorded on Discogs. \n \nMy average collection is worth $',discogs_df['_Median_'].mean().round(2))

## What is the most expensive record I bought?

In [ ]:
max_cost = discogs_df[discogs_df['Collection_Cost']==discogs_df['Collection_Cost'].max()]
max_cost

In [ ]:
max_cost[['Artist','Title','Format','Collection_Cost','_Median_','Genre','Subgenre']]

In [ ]:
print('This record has 7 full length albums and 3 EPs plus a DVD and a lyric book')

The most expensive record I own is the Meshuggah - 25 Years Of Musical Deviance, Boxset.

## What is the least expensive record I own?

In [ ]:
min_cost = discogs_df[discogs_df['Collection_Cost']==discogs_df['Collection_Cost'].min()]
min_cost

In [ ]:
min_cost[['Artist','Title','Format','Collection_Cost','_Median_','Genre','Subgenre']]

In [ ]:
print('I got this record for free from a friend')

## What is the least expensive record I bought?

In [ ]:
purchased_df = discogs_df[discogs_df.Collection_Cost > 0]
min_purchase = purchased_df[purchased_df['Collection_Cost']==purchased_df['Collection_Cost'].min()]
min_purchase[['Artist','Title','Format','Collection_Cost','_Median_','Genre','Subgenre']]

I bought Wyclef Jean 2xLP at the basement of Dusty Groove for 50 cents.

## What are the top 10 records I spent the most?

In [ ]:
discogs_df[['CollectionFolder','Record','Genre','Subgenre',
            'Collection_Autographed','Collection_Cost',
            '_Min_','_Median_','_Max_']].sort_values('Collection_Cost',ascending = False).head(10)

In [ ]:
discogs_df[['CollectionFolder','Record','Genre','Subgenre',
            'Collection_Autographed','Collection_Cost',
            '_Min_','_Median_','_Max_']].sort_values('Collection_Cost',ascending = False).groupby('CollectionFolder').head(10)

## What are the top 10 bands with the most expensive records purchased?

In [ ]:
record_sum   = discogs_df.groupby(['Artist2']).agg({
                       'Title': 'count',
                       'Label': 'nunique',
                       'Collection_Cost': 'sum',
                       '_Min_': 'sum',
                       '_Median_': 'sum',
                       '_Max_': 'sum'
                        }).reset_index().sort_values(by='_Median_',ascending=False)

## Plot top 10 bands that I purchased the most by cost

In [ ]:
chart = sns.catplot(x="Artist2", y="Collection_Cost", kind="bar", 
                    data=record_sum.sort_values('Collection_Cost',ascending=False).head(10),
                    height=6, aspect=1.5)
chart.set_xticklabels(rotation=45, horizontalalignment='right')
plt.title('Top 10 Most Spent Records by Bands')

record_sum.sort_values('Collection_Cost',ascending=False).head(10)

## What the most records purchased by the Top 10 bands?

In [ ]:
chart = sns.catplot(x="Artist2", y="Title", kind="bar", 
                    data=record_sum.sort_values('Title',ascending=False).head(10),
                    height=6, aspect=1.5)
chart.set_xticklabels(rotation=45, horizontalalignment='right')
plt.title('Top 10 Most Frequently Bought Records by Bands')

record_sum.sort_values('Title',ascending=False).head(10)

## What are the top 10 most valuable record by min?

In [ ]:
chart = sns.catplot(x="Record", y="_Min_", kind="bar", 
                    data=discogs_df.sort_values('_Max_',ascending=False).head(10),
                    height=6, aspect=1.5)
chart.set_xticklabels(rotation=45, horizontalalignment='right')
plt.title('Top 10 Most Valuable Records by Min')

discogs_df[['Record','Collection_Cost',
            '_Min_','_Median_','_Max_']].sort_values('_Min_',ascending=False).head(10)

## What are the top 10 most valuable record by median?

In [ ]:
chart = sns.catplot(x="Record", y="_Median_", kind="bar", 
                    data=discogs_df.sort_values('_Median_',ascending=False).head(10),
                    height=6, aspect=1.5)
chart.set_xticklabels(rotation=45, horizontalalignment='right')
plt.title('Top 10 Most Valuable Records by Median')

discogs_df[['Record','Title','Subgenre','Collection_Cost',
            '_Min_','_Median_','_Max_']].sort_values('_Median_',ascending=False).head(10)

## What are the top 10 most valuable record by max?

In [ ]:
chart = sns.catplot(x="Record", y="_Max_", kind="bar", 
                    data=discogs_df.sort_values('_Max_',ascending=False).head(10),
                    height=6, aspect=1.5)
chart.set_xticklabels(rotation=45, horizontalalignment='right')
plt.title('Top 10 Most Valuable Records by Max')

discogs_df[['Record','Collection_Cost',
            '_Min_','_Median_','_Max_']].sort_values('_Max_',ascending=False).head(10)

The Meshuggah boxset was sold for 514.17 and I bought it for 232.

## What are the Top 10 albums sold by average?

In [ ]:
record_mean   = discogs_df.groupby(['Artist2']).agg({
                       'Title': 'count',
                       'Label': 'nunique',
                       'Collection_Cost': 'mean',
                       '_Min_': 'mean',
                       '_Median_': 'mean',
                       '_Max_': 'mean'
                        }).reset_index().sort_values(by='_Median_',ascending=False)

In [ ]:
chart = sns.catplot(x="Artist2", y="_Median_", kind="bar", 
                    data=record_mean.sort_values('_Median_',ascending=False).head(10),
                    height=6, aspect=1.5)
plt.title('Top 10 Average Sold by Median')
chart.set_xticklabels(rotation=45, horizontalalignment='right')
record_mean.sort_values('_Median_',ascending=False).head(10)

## What are the Top 10 albums sold by max?

In [ ]:
chart = sns.catplot(x="Artist2", y="_Max_", kind="bar", 
                    data=record_mean.sort_values('_Max_',ascending=False).head(10),
                    height=6, aspect=1.5)
plt.title('Top 10 Average Sold by Max')
chart.set_xticklabels(rotation=45, horizontalalignment='right')
record_mean.sort_values('_Max_',ascending=False).head(10)

### Opeth Analysis

In [ ]:
discogs_df[['Artist2','Title','Collection_Cost','_Min_','_Median_','_Max_']][discogs_df.Artist2 == 'Opeth'].sort_values('_Max_',ascending=False)

# Label Analysis

In [ ]:
label_sum   = discogs_df.groupby(['Label']).agg({
                       'Artist2': 'nunique',
                       'Title': 'count',
                       'Collection_Cost': 'sum',
                       '_Min_': 'sum',
                       '_Median_': 'sum',
                       '_Max_': 'sum'
                        }).reset_index().sort_values(by='_Median_',ascending=False)
label_sum.head()

## Which record labels did I spend the most?

In [ ]:
chart = sns.catplot(x="Label", y="Collection_Cost", kind="bar", 
                    data=label_sum.sort_values('Collection_Cost',ascending=False).head(10),
                    height=6, aspect=1.5)
chart.set_xticklabels(rotation=45, horizontalalignment='right')
label_sum.sort_values('Collection_Cost',ascending=False).head(10)

In [ ]:
discogs_df.columns

In [ ]:
discogs_df.Date_Added = pd.to_datetime(discogs_df.Date_Added).dt.strftime('%m/%Y')
discogs_df.Date_Added.head()

# Time Series

### Group data by `Released` and `Date_Added` then get sum by `Collection_Cost`, `_Min_`, `_Median_`, and `_Max_`.

In [ ]:
time_df  = discogs_df.groupby(['Released','Date_Added']).agg({
                               'Collection_Cost': 'sum',
                               'Artist': 'nunique',
                               'Artist2': 'nunique',
                               'Title': 'count',
                               '_Min_': 'sum',
                               '_Median_': 'sum',
                               '_Max_': 'sum'
                                }).reset_index().sort_values(by='Date_Added',ascending=True)
time_df.head()

Plot `Date_Added` over `Collection_Cost`

In [ ]:
plt.plot(time_df['Date_Added'], time_df['Collection_Cost'])
plt.gcf().autofmt_xdate()
plt.show()

Plot `Date_Added` over `Title`

In [ ]:
plt.plot(time_df['Date_Added'], time_df['Title'])
plt.gcf().autofmt_xdate()
plt.show()

Plot `Date_Added` over `Artist2`

There were some days that I logged many records at once on discogs.

In [ ]:
plt.plot(time_df['Date_Added'], time_df['Artist2'])
plt.gcf().autofmt_xdate()
plt.show()

### Group data by `Released`

In [ ]:
year_df  = discogs_df.groupby(['Released']).agg({
                               'Artist2': 'count',
                               'Title': 'count',
                               'CollectionFolder':'count',
                               'Genre':'count', 
                               'Subgenre':'count',
                               'Collection_Cost': 'sum',
                               '_Min_': 'sum',
                               '_Median_': 'sum',
                               '_Max_': 'sum'
                                }).reset_index().sort_values(by='Released',ascending=True)
year_df.head()

In [ ]:
year_df['Released'].describe()

In [ ]:
plt.plot(year_df['Released'], year_df['Artist2'])
plt.gcf().autofmt_xdate()
plt.show()

In [ ]:
plt.plot(year_df['Released'], year_df['Title'])
plt.gcf().autofmt_xdate()
plt.show()

In [ ]:
chart = sns.catplot(x="Released", y="Title", kind="bar", 
                    data=year_df.sort_values('Released',ascending=True),
                    height=6, aspect=1.5)
chart.set_xticklabels(rotation=45, horizontalalignment='right')

In [ ]:
chart = sns.catplot(x="Released", y="_Min_", kind="bar", 
                    data=year_df.sort_values('Released',ascending=True),
                    height=6, aspect=1.5)
chart.set_xticklabels(rotation=45, horizontalalignment='right')

In [ ]:
chart = sns.catplot(x="Released", y="_Median_", kind="bar", 
                    data=year_df.sort_values('Released',ascending=True),
                    height=6, aspect=1.5)
chart.set_xticklabels(rotation=45, horizontalalignment='right')

In [ ]:
chart = sns.catplot(x="Released", y="_Max_", kind="bar", 
                    data=year_df.sort_values('Released',ascending=True),
                    height=6, aspect=1.5)
chart.set_xticklabels(rotation=45, horizontalalignment='right')

In [ ]:
chart = sns.catplot(x="Released", y="Collection_Cost", kind="bar", 
                    data=year_df.sort_values('Released',ascending=True),
                    height=6, aspect=1.5)
chart.set_xticklabels(rotation=45, horizontalalignment='right')

In [ ]:
# https://seaborn.pydata.org/generated/seaborn.FacetGrid.html
# bins = np.arange(0, 65, 5)
# g = sns.FacetGrid(discogs_df, col="Format",  row="Collection_Autographed")
# g = g.map(plt.hist, "Collection_Cost", bins=bins, color="r")

In [ ]:
# tips = sns.load_dataset("tips")
# g = sns.FacetGrid(tips, col="time",  row="smoker")
# g = g.map(plt.hist, "total_bill")

# Genre and Subgenre analysis

In [ ]:
discogs_df.Genre.value_counts()

In [ ]:
fig, axes = plt.subplots(2, 2)

ax = sns.boxplot(x="Genre", y="Collection_Cost", data=discogs_df, orient='v', 
    ax=axes[0, 0])
ax = sns.boxplot(x="Genre", y="_Min_", data=discogs_df, orient='v', 
    ax=axes[0, 1])
ax = sns.boxplot(x="Genre", y="_Median_", data=discogs_df, orient='v', 
    ax=axes[1, 0])
ax = sns.boxplot(x="Genre", y="_Max_", data=discogs_df, orient='v', 
    ax=axes[1, 1])

Remove records less than 2

In [ ]:
genre_reduced = discogs_df['Genre'].value_counts()[lambda x: x>2].index.tolist()
genre_reduced

In [ ]:
genre_reduced_df = discogs_df[(discogs_df['Genre'].isin(genre_reduced)) & (discogs_df.Collection_Cost < 200)]
genre_reduced_df.head()

In [ ]:
genre_reduced_df.Genre.value_counts()

## Boxplot Genre by Collection_Cost, Min, Median, and Max

In [ ]:
fig, axes = plt.subplots(2, 2)

ax = sns.boxplot(x="Genre", y="Collection_Cost", data=genre_reduced_df, orient='v', 
    ax=axes[0, 0])
ax = sns.boxplot(x="Genre", y="_Min_", data=genre_reduced_df, orient='v', 
    ax=axes[0, 1])
ax = sns.boxplot(x="Genre", y="_Median_", data=genre_reduced_df, orient='v', 
    ax=axes[1, 0])
ax = sns.boxplot(x="Genre", y="_Max_", data=genre_reduced_df, orient='v', 
    ax=axes[1, 1])

In [ ]:
sns.set_style('whitegrid')
sns.pairplot(genre_reduced_df[['Genre','Collection_Cost','_Min_','_Median_','_Max_']],hue='Genre',size=3)
plt.show()

In [ ]:
sns.FacetGrid(genre_reduced_df,hue="Genre",size=8).map(sns.distplot,
'Collection_Cost').add_legend()
plt.show()

In [ ]:
sns.FacetGrid(genre_reduced_df,hue="Genre",size=8).map(sns.distplot,
'_Min_').add_legend()
plt.show()

In [ ]:
sns.FacetGrid(genre_reduced_df,hue="Genre",size=8).map(sns.distplot,
'_Median_').add_legend()
plt.show()

In [ ]:
sns.FacetGrid(genre_reduced_df,hue="Genre",size=8).map(sns.distplot,
'_Max_').add_legend()
plt.show()

In [ ]:
sns.pairplot(genre_reduced_df, x_vars=["Collection_Cost"], y_vars=['_Min_','_Median_','_Max_'],
             hue="Genre", height=5, aspect=1.5, kind="reg");

## Subgenre Analysis

In [ ]:
discogs_df.Subgenre.value_counts()

In [ ]:
subgenre_reduced = discogs_df['Subgenre'].value_counts()[lambda x: x>5].index.tolist()
subgenre_reduced

In [ ]:
subgenre_reduced_df = discogs_df[(discogs_df['Subgenre'].isin(subgenre_reduced)) & (discogs_df.Collection_Cost < 200)]
subgenre_reduced_df.head()

## Boxplot Subgenre by Collection_Cost, Min, Median, and Max

In [ ]:
fig, axes = plt.subplots(2, 2)

ax = sns.boxplot(x="Subgenre", y="Collection_Cost", data=subgenre_reduced_df, orient='v', 
    ax=axes[0, 0])
ax.set_xticklabels(ax.get_xticklabels(),rotation=45)
ax = sns.boxplot(x="Subgenre", y="_Min_", data=subgenre_reduced_df, orient='v', 
    ax=axes[0, 1])
ax.set_xticklabels(ax.get_xticklabels(),rotation=45)
ax = sns.boxplot(x="Subgenre", y="_Median_", data=subgenre_reduced_df, orient='v', 
    ax=axes[1, 0])
ax.set_xticklabels(ax.get_xticklabels(),rotation=45)
ax = sns.boxplot(x="Subgenre", y="_Max_", data=subgenre_reduced_df, orient='v', 
    ax=axes[1, 1])
ax.set_xticklabels(ax.get_xticklabels(),rotation=45)

Reduce to Top 5 subgenres

In [ ]:
subgenre_reduced1 = discogs_df['Subgenre'].value_counts()[lambda x: x>25].index.tolist()
subgenre_reduced_df1 = discogs_df[(discogs_df['Subgenre'].isin(subgenre_reduced1)) & (discogs_df.Collection_Cost < 200)]

In [ ]:
subgenre_reduced_df1.head()

## Boxplot Reduced Subgenre by Collection_Cost, Min, Median, and Max

In [ ]:
fig, axes = plt.subplots(2, 2)

ax = sns.boxplot(x="Subgenre", y="Collection_Cost", data=subgenre_reduced_df1, orient='v', 
    ax=axes[0, 0])
ax.set_xticklabels(ax.get_xticklabels(),rotation=45)
ax = sns.boxplot(x="Subgenre", y="_Min_", data=subgenre_reduced_df1, orient='v', 
    ax=axes[0, 1])
ax.set_xticklabels(ax.get_xticklabels(),rotation=45)
ax = sns.boxplot(x="Subgenre", y="_Median_", data=subgenre_reduced_df1, orient='v', 
    ax=axes[1, 0])
ax.set_xticklabels(ax.get_xticklabels(),rotation=45)
ax = sns.boxplot(x="Subgenre", y="_Max_", data=subgenre_reduced_df1, orient='v', 
    ax=axes[1, 1])
ax.set_xticklabels(ax.get_xticklabels(),rotation=45)

## 3D Plot of Collection_Cost, Median, and Max

In [ ]:
import plotly.express as px
fig = px.scatter_3d(subgenre_reduced_df1, x='Collection_Cost', y='_Median_', z='_Max_', color='Subgenre')
fig.show()

## Plot Collection_Cost vs Median by Subgenre

In [ ]:
sns.set_style('whitegrid');
sns.FacetGrid(subgenre_reduced_df1,hue='Subgenre',size=10).map(plt.scatter,
'Collection_Cost','_Median_').add_legend();
plt.show();

## Plot Collection_Cost, Min, Median, and Max by Subgenre

In [ ]:
sns.set_style('whitegrid')
sns.pairplot(subgenre_reduced_df1[['Subgenre','Collection_Cost','_Min_','_Median_','_Max_']],hue='Subgenre',size=3)
plt.show()

In [ ]:
sns.pairplot(subgenre_reduced_df1, x_vars=["Collection_Cost"], y_vars=['_Min_','_Median_','_Max_'],
             hue="Subgenre", height=5, aspect=1.5, kind="reg");

## Distribution Plot Collection Cost by Subgenre

In [ ]:
sns.FacetGrid(subgenre_reduced_df1,hue="Subgenre",size=8).map(sns.distplot,
'Collection_Cost').add_legend()
plt.show()

## Distribution plot Min by Subgenre

In [ ]:
sns.FacetGrid(subgenre_reduced_df1,hue="Subgenre",size=8).map(sns.distplot,
'_Min_').add_legend()
plt.show()

## Distribution Plot Median by Subgenre

In [ ]:
sns.FacetGrid(subgenre_reduced_df1,hue="Subgenre",size=8).map(sns.distplot,
'_Median_').add_legend()
plt.show()

## Distribution Plot Max by Subgenre

In [ ]:
sns.FacetGrid(subgenre_reduced_df1,hue="Subgenre",size=8).map(sns.distplot,
'_Max_').add_legend()
plt.show()

# Format Analysis

### What are my top 10 valuable records by format?

In [ ]:
format_df = discogs_df[(discogs_df.CollectionFolder != 'Tape') & (discogs_df.Collection_Cost < 200)]

In [ ]:
format_df1 = discogs_df[(discogs_df.CollectionFolder != 'Tape')]
format_top20_df = format_df1.sort_values("_Median_",ascending = False).groupby('CollectionFolder').head(20)

format_top20_df[['CollectionFolder','Record','Genre','Subgenre',
                 'Collection_Autographed','Collection_Cost','_Median_']]

In [ ]:
sns.FacetGrid(format_df,hue="CollectionFolder",size=8).map(sns.distplot,'Collection_Cost').add_legend()
plt.show()

In [ ]:
sns.FacetGrid(format_df,hue="CollectionFolder",size=8).map(sns.distplot,'_Min_').add_legend()
plt.show()

In [ ]:
sns.FacetGrid(format_df,hue="CollectionFolder",size=8).map(sns.distplot,'_Median_').add_legend()
plt.show()

In [ ]:
sns.FacetGrid(format_df,hue="CollectionFolder",size=8).map(sns.distplot,'_Max_').add_legend()
plt.show()

In [ ]:
sns.set_style('whitegrid')
sns.pairplot(format_df[['CollectionFolder','Collection_Cost','_Min_','_Median_','_Max_']],hue='CollectionFolder',size=3)
plt.show()

In [ ]:
fig, axes = plt.subplots(2, 2)

ax = sns.boxplot(x="CollectionFolder", y="Collection_Cost", data=format_df, orient='v', 
    ax=axes[0, 0])
ax = sns.boxplot(x="CollectionFolder", y="_Min_", data=format_df, orient='v', 
    ax=axes[0, 1])
ax = sns.boxplot(x="CollectionFolder", y="_Median_", data=format_df, orient='v', 
    ax=axes[1, 0])
ax = sns.boxplot(x="CollectionFolder", y="_Max_", data=format_df, orient='v', 
    ax=axes[1, 1])

# Autographed Analysis

In [ ]:
auto_df = discogs_df[(discogs_df.Collection_Autographed == 'Yes')]
auto_top20_df = auto_df.sort_values("_Max_",ascending = False).groupby('CollectionFolder').head(10)

auto_top20_df[['CollectionFolder','Record','Genre','Subgenre',
                 'Collection_Autographed','Collection_Cost','_Min_','_Median_','_Max_']]

In [ ]:
sns.FacetGrid(auto_df,hue="CollectionFolder",size=8).map(sns.distplot,'Collection_Cost').add_legend()
plt.show()

In [ ]:
discogs_df1 = discogs_df[discogs_df.Collection_Cost < 200]

sns.set_style('whitegrid')
sns.pairplot(discogs_df1[['Collection_Autographed','Collection_Cost','_Min_','_Median_','_Max_']],hue='Collection_Autographed',size=3)
plt.show()

In [ ]:
fig, axes = plt.subplots(2, 2)

ax = sns.boxplot(x="Collection_Autographed", y="Collection_Cost", data=discogs_df1, orient='v', 
    ax=axes[0, 0])
ax = sns.boxplot(x="Collection_Autographed", y="_Min_", data=discogs_df1, orient='v', 
    ax=axes[0, 1])
ax = sns.boxplot(x="Collection_Autographed", y="_Median_", data=discogs_df1, orient='v', 
    ax=axes[1, 0])
ax = sns.boxplot(x="Collection_Autographed", y="_Max_", data=discogs_df1, orient='v', 
    ax=axes[1, 1])

In [ ]:
import plotly.express as px
fig = px.scatter(subgenre_reduced_df1, x="Collection_Cost", y="_Median_", color='Collection_Autographed', 
                 facet_col="Subgenre",hover_name="Subgenre",hover_data=["Artist2","Title","Record"])
fig.update_xaxes(matches=None)
fig.show()

In [ ]:
fig = px.scatter(subgenre_reduced_df1[subgenre_reduced_df1.CollectionFolder != 'Tape'], 
                 x="Collection_Cost", y="_Median_", color='Collection_Autographed', 
                 facet_col="Subgenre",facet_row="CollectionFolder",
                 hover_name="Subgenre",hover_data=["Artist2","Title","Record"])
fig.update_xaxes(matches=None)
fig.show()

In [ ]:
year_sub_df  = subgenre_reduced_df1.groupby(['Released','Subgenre']).agg({
                                         'Artist2': 'count',
                                         'Title': 'count',
                                         'Collection_Cost': 'sum',
                                         '_Min_': 'sum',
                                         '_Median_': 'sum',
                                         '_Max_': 'sum'
                                         }).reset_index().sort_values(by='Released',ascending=True)

# Round
year_sub_df.head()

In [ ]:
year_sub_cost_df = year_sub_df.pivot("Released", "Subgenre", "Collection_Cost")
year_sub_cost_df.round(0).tail()

In [ ]:
ax = sns.heatmap(year_sub_cost_df, annot=True, fmt="")

In [ ]:
year_sub_title_df = year_sub_df.pivot("Released", "Subgenre", "Title")
year_sub_title_df.tail()

In [ ]:
ax = sns.heatmap(year_sub_title_df, annot=True, fmt="")

In [ ]:
year_sub_title_df1 = year_sub_df.pivot("Released", "Subgenre", "Title")

ax = sns.heatmap(year_sub_title_df1, annot=True, fmt="f")

In [ ]:
year_sub_title_df1 = year_sub_df[year_sub_df.Released >= 2010].pivot("Released", "Subgenre", "Title")
ax = sns.heatmap(year_sub_title_df1, annot=True, fmt="")

In [ ]:
year_genre_df  = discogs_df.groupby(['Released','Genre']).agg({
                                     'Artist2': 'count',
                                     'Title': 'count',
                                     'Collection_Cost': 'sum',
                                     '_Min_': 'sum',
                                     '_Median_': 'sum',
                                     '_Max_': 'sum'
                                    }).reset_index().sort_values(by='Released',ascending=True)
year_genre_df.head()

In [ ]:
year_genre_title_df1 = year_genre_df.pivot("Released", "Genre", "Title")
ax = sns.heatmap(year_genre_title_df1, annot=True, fmt="")

In [ ]:
year_genre_title_df1 = year_genre_df[year_genre_df.Released >= 2010].pivot("Released", "Genre", "Title")
ax = sns.heatmap(year_genre_title_df1, annot=True, fmt="")

## Create animation for year

###  Example

In [ ]:
df = px.data.gapminder()
df.head()

In [ ]:
fig = px.bar(df, x="continent", y="pop", color="continent",
  animation_frame="year", animation_group="country",
             range_y=[0,4000000000])
fig.show()

In [ ]:
fig = px.bar(df, y="continent", x="pop", color="continent",
  animation_frame="year", animation_group="country", orientation = 'h',
             range_x=[0,4000000000])
fig.show()

### Subgenre By Year

In [ ]:
subgenre_reduced_df.head()

In [ ]:
year_sub_df1  = subgenre_reduced_df.groupby(['Released','Subgenre']).agg({
                                         'Title': 'count',
                                         'Collection_Cost': 'sum',
                                         '_Min_': 'sum',
                                         '_Median_': 'sum',
                                         '_Max_': 'sum'
                                         }).reset_index().sort_values(by='Released',ascending=True)

# Round
year_sub_df1.head()

In [ ]:
year_sub_df1['Released'].unique()

In [ ]:
Released = pd.DataFrame({'Released':year_sub_df1['Released'].unique()})
Released

In [ ]:
year_sub_df1['Subgenre'].unique()

In [ ]:
Subgenre = pd.DataFrame({'Subgenre':year_sub_df1['Subgenre'].unique()})
Subgenre

In [ ]:
def cartesian_product_basic(left, right):
    return (
       left.assign(key=1).merge(right.assign(key=1), on='key').drop('key', 1))

In [ ]:
released_subgenre_df = cartesian_product_basic(Released, Subgenre)
released_subgenre_df.head()

Left Join released_subgenre_df on year_sub_df1

In [ ]:
year_sub_df1.head()

In [ ]:
subgenre_df2 = released_subgenre_df.merge(year_sub_df1, on = ['Released','Subgenre'],how='left').fillna(0).sort_values(by=['Released','Subgenre'],ascending=True)
subgenre_df2.head()

In [ ]:
subgenre_cum_df2 = subgenre_df2.groupby(['Released','Subgenre']).sum().groupby(['Subgenre']).cumsum().reset_index()
subgenre_cum_df2.head()

Start animations here

## Title vs. Subgenre by Year

In [ ]:
fig = px.bar(subgenre_cum_df2, x="Title", y="Subgenre", color="Subgenre",
  animation_frame="Released", range_x=[0,90], orientation = 'h')
fig.show()

## Collection_Cost vs. Subgenre by Year

In [ ]:
fig = px.bar(subgenre_cum_df2, x="Collection_Cost", y="Subgenre", color="Subgenre",
  animation_frame="Released", range_x=[0,1600], orientation = 'h')
fig.show()

## Median vs. Subgenre by Year

In [ ]:
fig = px.bar(subgenre_cum_df2, x="_Median_", y="Subgenre", color="Subgenre",
  animation_frame="Released", range_x=[0,2000], orientation = 'h')
fig.show()

## Title vs. Collection Cost by Year

In [ ]:
px.scatter(subgenre_cum_df2,x = 'Title', y = 'Collection_Cost',color='Subgenre',
               size='_Median_',animation_frame='Released', animation_group='Subgenre',range_x = [0,100],
              range_y = [0,2000],text='Subgenre')

## Scatterplot - Title vs. Collection Cost by Year

In [ ]:
px.scatter(subgenre_cum_df2,x = 'Collection_Cost', y = '_Median_',color='Subgenre',
               size='Title',animation_frame='Released', animation_group='Subgenre',range_x = [0,2000],
              range_y = [0,2100],text='Subgenre')

Export data for Floursh

In [ ]:
subgenre_cum_df2 = subgenre_cum_df2.sort_values(by=['Released','Collection_Cost'],ascending=[True,False])
subgenre_cum_df2.to_csv('../output/discogs_cum_sum_data.csv',index = False)

In [ ]:
subgenre_cum_df2.set_index(['Released','Subgenre']).T